### PR TITLE
FFT + Time Reversal Reconstruction

### PR DESCRIPTION
**Pull request by faberno**
This PR implements the FFT reconstructions (kspaceLineRecon, kspacePlaneRecon) and fixes Time Reversal.

Added examples: 
- pr_2D_FFT_line_sensor
- pr_3D_FFT_plane_sensor
- pr_2D_TR_line_sensor
- pr_3D_TR_planar_sensor

This PR is just a draft and not ready for review yet!

TODO:
- [x] Align style of examples
- [x] Add Notebooks + Colab
- [x] Add documentation
- [x] Add tests

### CODE REVIEW

**Comment by waltsims** [comment]
Hey @faberno,

just pining you on this. Is this something you would like to continue to work on?

Best,
Walter

**Comment by faberno** [comment]
Hey @waltsims,

sorry for the inactivity, next week I will have time to continue it :)

Best,
Fabian

**Comment by faberno** [comment]
Hey @waltsims,
could you review my changes?

File: kwave/executor.py
Comment (faberno) [inline kwave/executor.py:1]: I added this PML cropping section to the run_simulation function, to have access to the simulation_options. Would you prefer it in the parse_executable_output? (At least, thats where you added an outcommented version of this.)

**Comment by waltsims** [comment]
Hey @faberno 

Thanks for this work! Before I review your changes, can you make sure your PR passed the CI? It looks like there are some failing tests. Also please be sure to [install precommit](https://k-wave-python.readthedocs.io/en/latest/development/development_environment.html) to ensure your PR is properly formatted

File: kwave/executor.py
Comment (faberno) [reply kwave/executor.py:1]: Also right now this change breaks the executor test, I will update it when the code is final.

**Comment by waltsims** [comment]
Hey @faberno,

Thanks again for the PR. I'll have a look at this this weekend.

File: kwave/executor.py
Comment (waltsims) [reply kwave/executor.py:1]: it would be great if you could make this a private subfunction `_remove_pml(...)` and place it in `parse_executable output` for now. That will make future refactoring easier.

**Review by waltsims — APPROVED** [review]
great examples. Minor comment. Do you still want to add further tests?

**Comment by waltsims** [comment]
Please resolve the conflict in the tests and we can merge this in unless you would like to make further changes.

File: kwave/executor.py
Comment (faberno) [reply kwave/executor.py:1]: Good idea! I just have the problem, that if `_remove_pml(...)` is inside `parse_executable_output` I can't test anymore if the pml is removed on the mock_dict (see: _test_sensor_data_cropping_with_pml_outside_), because the whole `parse_executable_output` is patched.

If I create a second private subfunction in `parse_executable_output` called `_load_sensor_data`, is there a way to only patch this instead? I didn't find a way to do this.

Or do you see another workaround? :D

**Comment by faberno** [comment]
Are there any more tests you would like to have? Otherwise it's fine for me

**Comment by faberno** [comment]
Hey @waltsims,

I added a last test for the coverage. I also made the _remove_pml(...) function a normal member function of Executor (instead of a subfunction in parse_executable output), to still be able to test the cropping. If thats alright for you, everything should be ready :)

**Comment by waltsims** [comment]
On hold until the license is updated.

**Comment by waltsims** [comment]
@faberno can you please update this PR with the new license file from master?

**Comment by waltsims** [comment]
@faberno can you please resolve the conflicts before I merge this in under the new LGPL-3.0 license?

**Comment by faberno** [comment]
@waltsims Done!

**Pull request merged**
